File: C++/station.cpp
//#include <iostream>
//#include <math.h>
//
//using namespace std;
//
//double x, y, z;
//double result = 100000000000, answer=100000000001, len=100000000000;
//double s1[3], s2[3];
//double vec[4][3];
//
//double getDistance(double pos1[3], double pos2[3]) {
//    len = (pos1[0] - pos2[0]) * (pos1[0] - pos2[0]);
//    len += (pos1[1] - pos2[1]) * (pos1[1] - pos2[1]);
//    len += (pos1[2] - pos2[2]) * (pos1[2] - pos2[2]);
//    return len;
//}
//
//void input() {
//    for(int i=0; i<4; i++) {
//        cin >> x >> y >> z;
//        vec[i][0] = x;
//        vec[i][1] = y;
//        vec[i][2] = z;
//    }
//}
//
//void solve() {
//    for(int k=0; k<3; k++) {
//        s1[k] = vec[0][k];
//    }
//    double lengthOfAB = sqrt(getDistance(vec[0], vec[1]));
//    double lengthOfCD = sqrt(getDistance(vec[2], vec[3]));
//
//    for(int i=0; i < lengthOfAB; i++) {
//        if (result < answer) answer = result;
//        else break;
//        s2[0] = vec[2][0];
//        s2[1] = vec[2][1];
//        s2[2] = vec[2][2];
//        for(int j=0; j < lengthOfCD; j++) {
//            len = getDistance(s1, s2);
//            if(len <= result) result = len;
//            s2[0] += ((vec[3][0] - vec[2][0]) / lengthOfCD);
//            s2[1] += ((vec[3][1] - vec[2][1]) / lengthOfCD);
//            s2[2] += ((vec[3][2] - vec[2][2]) / lengthOfCD);
//        }
//        s1[0] += ((vec[1][0] - vec[0][0]) / lengthOfAB);
//        s1[1] += ((vec[1][1] - vec[0][1]) / lengthOfAB);
//        s1[2] += ((vec[1][2] - vec[0][2]) / lengthOfAB);
//    }
//    cout << round(sqrt(answer)) << endl;
//}
//
//int main() {
//    input();
//    solve();
//    return 0;
//}

File: python/조고리즘/station.py
import math
import numpy as np
def getDistanceFromDotToLine(startPos, endPos , Dot):
    vec1 = [ai - bi for ai, bi in zip(startPos, Dot)]
    vec2 = [ci - di for ci, di in zip(endPos, Dot)]
    area = getLengthOfLine(np.cross(vec1,vec2))
    print("area: ",area)
    AB = getDistance(startPos,endPos)
    distance = area / AB
    print("distance: ",distance)
    return distance
def getDistance(pos1, pos2):
    return math.sqrt(sum([(pos1i-pos2i)**2 for pos1i, pos2i in zip(pos1,pos2)]))
def getLengthOfLine(vector):
    return math.sqrt(vector[0]**2 + vector[1]**2 + vector[2]**2)
def innerProduct(vec1,vec2):
    return sum([vec1i*vec2i for vec1i, vec2i in zip(vec1,vec2)])
def getAngle(vec1, vec2):
    return abs(np.dot(vec1,vec2)) / (getLengthOfLine(vec1) * getLengthOfLine(vec2))
def getLineEquation(pos, directionVector):
    return [[i,j] for i, j in zip(pos, directionVector)]
def isMeet(abEquation,cdEquation):
    A=np.array([[1,1],[2,4]])
    B=np.array([cdEquation[0][0] - abEquation[0][0]],[cdEquation[1][0] - abEquation[1][0]],[cdEquation[2][0] - abEquation[2][0]])
    print(B)
    try:
        np.linalg.solve(A, B)
        return True
    except:
        return False
def main():
    aPos = np.array(list(map(float, input().split())))
    bPos = np.array(list(map(float, input().split())))
    cPos = np.array(list(map(float, input().split())))
    dPos = np.array(list(map(float, input().split())))

    vec1 = [ai - bi for ai, bi in zip(aPos, bPos)]
    vec2 = [ci - di for ci, di in zip(cPos, dPos)]

    print(vec1,vec2)

    abEquation = getLineEquation(aPos, vec1)
    cdEquation = getLineEquation(cPos, vec2)

    print("Equation: ", abEquation,cdEquation)

    angle = getAngle(vec1, vec2)

    # if isMeet(abEquation, cdEquation):
    #     print("교점을 가짐")
    #     print(0)
    #     return
    if round(angle) == 1:
        print("평행")
        print(getDistanceFromDotToLine(aPos,bPos,cPos))
        return

main()








